### PR TITLE
Pov edit

### PR DESCRIPTION
**Pull request by vegetable68**
This PR aims to 

- Put the MediaWiki parsing the sentence tokenization step into a sandbox.
- Limit the memory usage and the execution time of the process in the sandbox.
- Add heuristic to skip revision pairs in order to find sentence revises more efficiently.

### CODE REVIEW

File: extract_pov_edits/dataflow_main.py [hunk 1]
```diff
@@ -90,6 +94,14 @@
 from apache_beam.io import filesystems
 from datetime import datetime
 import nltk
+import resource
+from threading import Timer
+import signal
+
+
+MEMLIMIT = 1024 * 1024 * 1024
+my_timeout = 2
```
Comment (sorensenjs) [inline extract_pov_edits/dataflow_main.py]: inconsistent naming of constants

File: extract_pov_edits/dataflow_main.py [hunk 2]
```diff
@@ -156,18 +182,23 @@ def process(self, element, bucket):
...
-    # Decompress the data dump
-    chunk_name = element
+    if isinstance(element, basestring):
+      chunk_name = element
+      page_ids = None
+    else:
+      # If specified on selected pages.
+      (chunk_name, data) = element
+      if data['datadump'] == []:
+         logging.info('USERLOG: chunk %s skipped.' % chunk_name)
```
Comment (sorensenjs) [inline extract_pov_edits/dataflow_main.py]: don't use % in logging statement:
 logging.info('USERLOG: chunk %s skipped.', chunk_name)

File: extract_pov_edits/dataflow_main.py [hunk 3]
```diff
@@ -178,48 +209,97 @@ def process(self, element, bucket, ingestFrom):
...
-        for sent in inserts:
-          ret['content'] = sent[0]
-          yield beam.pvalue.TaggedOutput('npov_improvements', json.dumps(ret))
-        for sent in context_equals:
-          ret['content'] = sent
-          yield beam.pvalue.TaggedOutput('pov_non_rejects', json.dumps(ret))
-      logging.info('CHUNK {chunk}: revision {revid} ingested, time elapsed: {time}.'.format(chunk=chunk_name, revid=last_revision, time=time.time() - last_completed))
+      if (content["comment"] is not None and "POV" in content["comment"]) or all_edits:
+        yield json.dumps((last_revision, content))
+      logging.info('INGESTION_LOG: CHUNK {chunk}: revision {revid} ingested, time elapsed: {time}.'.format(
```
Comment (sorensenjs) [inline extract_pov_edits/dataflow_main.py]: don't use .format() in LOG statements.
logger supports passing dictionary of via extra
https://docs.python.org/2/library/logging.html#logging.Logger.debug

File: extract_pov_edits/dataflow_main.py [hunk 4]
```diff
@@ -178,48 +209,97 @@ def process(self, element, bucket, ingestFrom):
...
+          chunk=chunk_name, revid=content['rev_id'], time=time.time() - last_completed))
       last_completed = time.time()
     if ingestFrom != 'local': os.system("rm %s" % chunk_name)
-    logging.info('USERLOG: Ingestion on file %s complete! %s lines emitted, last_revision %s' % (chunk_name, i, last_revision))
+    logging.info('USERLOG: Ingestion on file %s complete! %s lines emitted' % (chunk_name, i))
```
Comment (sorensenjs) [inline extract_pov_edits/dataflow_main.py]: Yeah... all of these.

File: extract_pov_edits/dataflow_main.py [hunk 5]
```diff
@@ -178,48 +209,97 @@ def process(self, element, bucket, ingestFrom):
...
+    (former, content) = json.loads(element)
+    if isSimilar(former, content):
+      # Only focus on revision pairs that are similar to look for sentence
+      # revises.
+      logging.info('EDIT_PROCESS_LOG: revision {revid} started.'.format(revid=content['rev_id']))
+      self.processed_revision_pairs.inc()
+      # Add memory and time limit to edit processing to prevent pipeline
+      # crashes.
+
+      tf = tempfile.NamedTemporaryFile(delete=False)
```
Comment (sorensenjs) [inline extract_pov_edits/dataflow_main.py]: use a with statement?

File: extract_pov_edits/dataflow_main.py [hunk 6]
```diff
@@ -178,48 +209,97 @@ def process(self, element, bucket, ingestFrom):
...
+      json.dump((former, content), tf)
+      tf.close()
+      process_cmd = ['python2', '-m', 'ingest_utils.run_processor', '-i', tf.name]
+      try:
```
Comment (sorensenjs) [inline extract_pov_edits/dataflow_main.py]: pull as much of this as you can out to it's own function.

File: extract_pov_edits/experimental/match_modify.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+      # Fix data formats.
+      data = {key[2:] : val for key, val in input_data.iteritems()}
+      data['rev_id'] = int(data['rev_id'])
+      ret.append(data)
+  return ret
+
+def matched(sent1, sent2):
+  tokens1 = word_tokenize(sent1)
+  tokens2 = word_tokenize(sent2)
+  overlapping_rate1 = len(set(tokens1)&set(tokens2))/float(len(set(tokens1)))
```
Comment (sorensenjs) [inline extract_pov_edits/experimental/match_modify.py]: spaces around operators.
don't compute set(tokens1) twice.

File: extract_pov_edits/ingest_utils/process.py [hunk 8]
```diff
@@ -149,3 +197,38 @@ def process(content, cur_sents):
        error = True
   return diff(sents, cur_sents, content['rev_id']), error
 
+
+def matched(sent1, sent2):
```
Comment (sorensenjs) [inline extract_pov_edits/ingest_utils/process.py]: is this here twice?

File: extract_pov_edits/process_test.py [hunk 9]
```diff
@@ -28,34 +28,101 @@
...
+
+MEMLIMIT = 1024 * 1024 * 1024
+my_timeout = 2
+
+def set_memory_limit(soft, hard):
+  resource.setrlimit(resource.RLIMIT_AS, (soft, hard))
 
 class TestWikiIngester(unittest.TestCase):
-  def test_normal_revs(self):
+  #def test_normal_revs(self):
```
Comment (sorensenjs) [inline extract_pov_edits/process_test.py]: commented out code?

File: extract_pov_edits/process_test.py [hunk 10]
```diff
@@ -28,34 +28,101 @@
...
+    ans_file = path.join('ingest_utils', 'testdata', 'dummy_test_ans_pairs.json')
     anses = []
     with open(input_file, 'r') as f:
+      former = None
       for i, line in enumerate(f):
-          content = json.loads(line)
-          ans, error = process(content, cur_sents)
-          anses.append(ans)
+        content = json.loads(line)
+        tf = tempfile.NamedTemporaryFile(delete=False)
```
Comment (sorensenjs) [inline extract_pov_edits/process_test.py]: use a with?

File: extract_pov_edits/ingest_utils/process.py (hunk 8 above)
Comment (vegetable68) [reply extract_pov_edits/ingest_utils/process.py]: No, but the same function was implemented in another file as well.

**Pull request merged**